src/cdi_CPEloss/Tabular_CP_Eloss.hh

```diff
@@ -0,0 +1,115 @@
+ * \brief Derived rtt_cdi::CPEloss class for tabular eloss.                     
+ *        This class implements the interface found in cdi/CPEloss.hh for        
+ *        the case where CP energy loss data is in tabular form, stored 
+ *        in a file.                
+ *                                                                               
+ */
+//===========================================================================//
+
+class Tabular_CP_Eloss : public rtt_cdi::CPEloss {
+public:
+  typedef std::vector<double> sf_double;
+  typedef stdex::extents<stdex::dynamic_extent,
+                         std::experimental::dynamic_extent,
+                         std::experimental::dynamic_extent>
+      dynamic_extents_3;
+
+private:
+  const std::string filename;
+  std::ifstream file;
+
+  rtt_units::UnitSystem us;
+  rtt_units::PhysicalConstants pc_cgs;
+
+  uint32_t n_energy;          //!< Number of gridpoints in projectile energy
+  uint32_t n_density;         //!< Number of gridpoints in target density
+  uint32_t n_temperature;     //!< Number of gridpoints in target temperature
+  double d_log_energy;        //!< Log spacing of projectile energy gridpoints
+  double d_log_density;       //!< Log spacing of target density gridpoints
+  double d_log_temperature;   //!< Log spacing of target temperature gridpoints
+  double min_log_energy;      //!< Log of minimum projectile energy
+  double min_log_density;     //!< Log of minimum target density
+  double min_log_temperature; //!< Log of minimum target temperature
+  double min_energy;          //!< Minimum target energy
+  double max_energy;          //!< Maximum target energy
+  double min_density;         //!< Minimum target density
+  double max_density;         //!< Maximum target density
+  double min_temperature;     //!< Minimum target temperature
+  double max_temperature;     //!< Maximum target temperature
+  sf_double energies;         //!< Vector of energy gridpoints
+  sf_double densities;        //!< Vector of density gridpoints
+  sf_double temperatures;     //!< Vector of temperature gridpoints
+  // Note that after unit conversions, *_energy is really *_speed
+
+  // Storage for tabulated data
+  stdex::basic_mdspan<double, dynamic_extents_3, std::experimental::layout_left>
+      stopping_data;
+
+  // Utility for skipping lines
+  void skip_lines(uint32_t nlines);
+
+  // Utility for reading a line of an eloss file and as a vector of strings
+  std::vector<std::string> const read_line();
+
+public:
+  // Constructor
+  Tabular_CP_Eloss(std::string filename_in, rtt_cdi::CParticle target_in,
+                   rtt_cdi::CParticle projectile_in);
+
+  // >>> ACCESSORS
+
+  double getEloss(const double temperature, const double density,
+                  const double v0) const;
+
+  bool data_in_tabular_form() const { return true; }
+
+  inline std::string getDataFilename() const { return filename; }
```

Comment:
Can we safely return a reference to the filename?

Reply:
I think this would conflict with `getDataFilename` in `Analytic_CP_Eloss`, which currently returns `0` via `CPEloss`. If we remove `getDataFilename` everywhere except `Tabular_CP_Eloss`, I think returning a reference is fine because `filename` is now `const`.

Reply:
ok